cmdlets/sqlserver/New-MiLink.ps1

```diff
@@ -0,0 +1,318 @@
+            HelpMessage = 'Enter secondary availability group name')]
+        [string]$SecondaryAvailabilityGroup,
+
+        [Parameter(Mandatory = $true,
+            ParameterSetName = 'InteractiveParameterSet',
+            HelpMessage = 'Enter instance link name')]
+        [Parameter(Mandatory = $true,
+            ParameterSetName = 'NonInteractiveParameterSet',
+            HelpMessage = 'Enter instance link name')]
+        [string]$LinkName
+
+        # auth params?
+        
+            
+    )
+    Begin {
+        $interactiveMode = ($PsCmdlet.ParameterSetName -eq "InteractiveParameterSet")
+        #if ($interactiveMode) {
+        #    $miCredential = Get-Credential -Message "Enter your SQL Managed instance credentials in order to login"
+        #}
+        #else {
+        #    $miCredential = $ManagedInstanceCredential
+        #}
+        Write-Verbose "Interactive mode enabled - $interactiveMode"
+    }
+    Process {
+        $ErrorActionPreference = "Stop"
+
+        # All databases that will be replicated via the link must be in full recovery mode and have at least one backup. 
+        # Recovery model selected: 1 = FULL, 2 = BULK_LOGGED, 3 = SIMPLE
+        $queryGetRecoveryModel = "select recovery_model from sys.databases where name = N'$DatabaseName'"
+        $recoveryModel = Invoke-SqlCmd -Query $queryGetRecoveryModel -ServerInstance $SqlInstance
+        if ($recoveryModel -ne 1) {
+            Write-Verbose "Set recovery model to FULL for database $DatabaseName [started]"
+            Invoke-SqlCmd -Query "ALTER DATABASE [$DatabaseName] SET RECOVERY FULL" -ServerInstance $SqlInstance      
+            Write-Verbose "Set recovery model to FULL for database $DatabaseName [completed]"
+        }
+
+        $backupHistory = Get-SqlBackupHistory -ServerInstance $SqlInstance -DatabaseName $DatabaseName
+        if (!$backupHistory) {
+            $backupDiskPath = Read-Host -Prompt "Creating database backup, enter path"
+            Write-Verbose "Execute backup for all databases you want to replicate [started]"
+            Invoke-SqlCmd -Query "BACKUP DATABASE [$DatabaseName] TO DISK = N'$backupDiskPath'" -ServerInstance $SqlInstance
+            Write-Verbose "Execute backup for all databases you want to replicate [completed]"
+            Get-SqlBackupHistory -ServerInstance $SqlInstance -DatabaseName $DatabaseName
+        }
+
+        Write-Verbose "Get MI info ($ManagedInstanceName) [started]"
+        $instance = Get-AzSqlInstance -ResourceGroupName $ResourceGroupName -Name $ManagedInstanceName
+        $miFQDN = $instance.FullyQualifiedDomainName       
+        $SqlMiInstanceConnectionString = "tcp://${miFQDN}:5022;Server=[$ManagedInstanceName]"
+        Write-Verbose "Get MI info ($ManagedInstanceName) [completed]"
+
+        # Figure out if we're on a VM https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows#example-scenarios-for-usage
```

Comment:
I think you just need to fail if you detect we're not on the VM.

Reply:
i agree. I think I will add a `begin { }` block that validates the env state (i.e. checks whether Init- cmdlet was ran / server requirements are met)